kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt

```diff
@@ -308,4 +314,209 @@ class WorkflowNodeTest {
+        it.writeUtf8WithLength("state:$state")
+      }
+
+      override fun restoreState(snapshot: Snapshot): String = snapshot.bytes.parse {
+        it.readUtf8WithLength()
+            .removePrefix("state:")
+      }
+    }
+    val originalNode = WorkflowNode(
+        workflow.id(),
+        workflow,
+        initialInput = "initial input",
+        snapshot = null,
+        baseContext = Unconfined
+    )
+
+    assertEquals("initial input", originalNode.compose(workflow, "foo"))
+    val snapshot = originalNode.snapshot(workflow)
+    assertNotEquals(0, snapshot.bytes.size())
+
+    val restoredNode = WorkflowNode(
+        workflow.id(),
+        workflow,
+        // This input should be ignored, since snapshot is non-null.
+        initialInput = "new input",
+        snapshot = snapshot,
+        baseContext = Unconfined
+    )
+    assertEquals("initial input", restoredNode.compose(workflow, "foo"))
+  }
+
+  @Test fun snapshots_empty_withoutChildren() {
+    val workflow = object : Workflow<String, String, Nothing, String> {
+      override fun initialState(input: String): String = input
+
+      override fun compose(
+        input: String,
+        state: String,
+        context: WorkflowContext<String, Nothing>
+      ): String {
+        return state
+      }
+
+      override fun snapshotState(state: String): Snapshot = Snapshot.EMPTY
+
+      override fun restoreState(snapshot: Snapshot): String = "restored"
+    }
+    val originalNode = WorkflowNode(
+        workflow.id(),
+        workflow,
+        initialInput = "initial input",
+        snapshot = null,
+        baseContext = Unconfined
+    )
+
+    assertEquals("initial input", originalNode.compose(workflow, "foo"))
+    val snapshot = originalNode.snapshot(workflow)
+    assertNotEquals(0, snapshot.bytes.size())
+
+    val restoredNode = WorkflowNode(
+        workflow.id(),
+        workflow,
+        // This input should be ignored, since snapshot is non-null.
+        initialInput = "new input",
+        snapshot = snapshot,
+        baseContext = Unconfined
+    )
+    assertEquals("restored", restoredNode.compose(workflow, "foo"))
+  }
+
+  @Test fun snapshots_nonEmpty_withChildren() {
+    var restoredChildState: String? = null
+    var restoredParentState: String? = null
+    val childWorkflow = object : Workflow<String, String, Nothing, String> {
+      override fun initialState(input: String): String = input
+
+      override fun compose(
+        input: String,
+        state: String,
+        context: WorkflowContext<String, Nothing>
+      ): String {
+        return state
+      }
+
+      override fun snapshotState(state: String): Snapshot = Snapshot.write {
+        it.writeUtf8WithLength("child state:$state")
+      }
+
+      override fun restoreState(snapshot: Snapshot): String = snapshot.bytes.parse {
+        it.readUtf8WithLength()
+            .removePrefix("child state:")
+            .also { state -> restoredChildState = state }
```

Comment:
`restoredChildState = it.eadUtf8WithLength()...`?

Reply:
The primary purpose of this function is to return the restored state, I feel like `also` expresses that this is a side-effect.